File: csrc/plugin/register_ops.cpp
/**
 * Copyright (c) Huawei Technologies Co., Ltd. 2025-2025. All rights reserved.
 * MindIE is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 */

#include <torch/library.h>

#include "rope.h"
#include "la.h"
#include "adalayernorm.h"
#include "la_preprocess.h"
#include "rainfusionattention.h"
#include "block_sparse_attention.h"
#include "sparse_block_estimate.h"
#include "layernorm.h"


TORCH_LIBRARY(mindiesd, m)
{
    m.def("rope(Tensor query, Tensor key, Tensor value, int mode) -> Tensor");
    m.def(
        "la(Tensor query, Tensor key, Tensor value, \
        Tensor? atten_mask=None, Tensor? alibi_mask=None, Tensor? \
        drop_mask=None, float scale_value=1.0, int head_num=2, str input_layout='BNSD', \
        float keep_prob=1.0, int pre_tokens=2147483647, int next_tokens=1, \
        bool is_highPrecision=True)  -> (Tensor, Tensor)");
    m.def("adaln(Tensor x, Tensor scale, Tensor shift, Tensor? weight=None, \
        Tensor? bias=None, float? epsilon=1e-5) \
        -> Tensor");
    m.def("adaln_v2(Tensor x, Tensor scale, Tensor shift, Tensor? weight=None, \
        Tensor? bias=None, float? epsilon=1e-5) \
        -> (Tensor, Tensor, Tensor)");
    m.def("la_preprocess(Tensor query, Tensor key, Tensor value, int align_len=256) \
        -> (Tensor, Tensor, Tensor)");
    m.def("rainfusionattention(Tensor query, Tensor key, Tensor value, Tensor select_idx, \
        Tensor select_num_idx, int[] blockshape, Tensor? attn_mask=None, int[]? actual_seq_qlen=None, \
        int[]? actual_seq_kvlen=None, Tensor? block_table=None, str q_input_layout='TND', str kv_input_layout='TND', \
        int head_num=1, int mask_type=0, float scale=1.0, \
        int inner_precise=1, int block_size=0) -> (Tensor, Tensor)");
    m.def("block_sparse_attention(Tensor query, Tensor key,  \
        Tensor value, Tensor sparse_mask, Tensor sparse_count_table,  \
        str input_layout='BNSD', int sparse_size=128, int num_heads=1, \
        int num_key_value_heads=1, float scale_value=1,  \
        bool causal=True, int inner_precise=1, int pre_tokens=214748647, int next_tokens=0, \
        int[]? actual_seq_lengths=None, int[]? actual_seq_lengths_kv=None)   \
        -> Tensor");
    m.def("sparse_block_estimate(Tensor query, Tensor key,  \
        int[]? actual_seq_lengths=None, int[]? actual_seq_lengths_kv=None,  \
        str input_layout='BNSD', int stride=8, int sparse_size=128,  \
        int num_heads=1, int num_key_value_heads=1, float scale_value=1,  \
        float threshold=1, bool causal=True, bool keep_sink=True,  \
        bool keep_recent=True, float row_sparse=1) \
        -> (Tensor, Tensor)");
    m.def("layernorm(Tensor input, int[] normalized_shape, Tensor? weight=None, Tensor? bias=None, float eps=1e-05, \
        int impl_mode=0) -> (Tensor, Tensor, Tensor)");
}


TORCH_LIBRARY_IMPL(mindiesd, PrivateUse1, m)
{
    m.impl("rope", &rope_mindie_sd_impl_npu);
    m.impl("la", &la_mindie_sd_impl_npu);
    m.impl("adaln", &adaln_mindie_sd_impl_npu);
    m.impl("adaln_v2", &adaln_v2_mindie_sd_impl_npu);
    m.impl("la_preprocess", &la_preprocess_mindie_sd_impl_npu);
    m.impl("rainfusionattention", &rainfusionattention_mindie_sd_impl_npu);
    m.impl("block_sparse_attention", &block_sparse_attention_impl_npu);
    m.impl("sparse_block_estimate", &sparse_block_estimate_mindie_sd_impl_npu);
    m.impl("layernorm", &layernorm_mindie_sd_impl_npu);
}

File: mindiesd/layers/register_ops.py
#!/usr/bin/env python
# coding=utf-8
# Copyright (c) Huawei Technologies Co., Ltd. 2025-2025. All rights reserved.
# MindIE is licensed under Mulan PSL v2.
# You can use this software according to the terms and conditions of the Mulan PSL v2.
# You may obtain a copy of Mulan PSL v2 at:
#          http://license.coscl.org.cn/MulanPSL2
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
# EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
# MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
# See the Mulan PSL v2 for more details.
from pathlib import Path
from functools import wraps
import os
from typing import Dict, Callable
import torch
from torch.library import Library
from ..utils import file_utils, ParametersInvalid, is_npu_available


MINDIE_NS = "mindiesd"  # 固定命名空间，与 torch.ops.mindiesd 对应


def _load_mindie_ops_library() -> None:
    """Load the MindIE custom operator shared library.
    
    Raises:
        ParametersInvalid: If the parent directory level is insufficient.
        FileNotFoundError: If the operator SO file is not found.
        PermissionError: If the SO file has invalid permissions.
    """
    current_path = Path(__file__).resolve()
    if len(current_path.parents) < 2:
        raise ParametersInvalid("Insufficient parent directory levels to locate plugin folder.")
    
    ops_path = current_path.parents[1] / "plugin"
    ops_path = file_utils.standardize_path(str(ops_path))
    ops_file = os.path.join(ops_path, "libPTAExtensionOPS.so")
    
    file_utils.check_file_safety(
        ops_file,
        permission_mode=file_utils.BINARY_FILE_PERMISSION
    )
    torch.ops.load_library(ops_file)


if is_npu_available():
    _load_mindie_ops_library()


def check_mindie_operator_exists(op_name: str) -> bool:
    """Check if a MindIE operator is registered in PyTorch.
    
    Args:
        op_name: Full name of the operator (e.g. "rope", "la")
    
    Returns:
        True if the operator exists, False otherwise.
    """
    try:
        getattr(torch.ops.mindiesd, op_name)
        return True
    except AttributeError:
        return False


if torch.__version__.startswith("2.1"):
    # PyTorch 2.1 使用 Library.impl
    _lib = Library(MINDIE_NS, "IMPL")
    
    def _compatible_register_fake(op_name: str):
        """Compatibility wrapper for PyTorch 2.1 fake registration."""
        def decorator(fake_func: Callable):
            @wraps(fake_func)
            def wrapper(*args, **kwargs):
                # Ensure all tensor inputs are on Meta device (required for PyTorch 2.1)
                args = [
                    a.to(device="meta") if isinstance(a, torch.Tensor) else a
                    for a in args
                ]
                kwargs = {
                    k: v.to(device="meta") if isinstance(v, torch.Tensor) else v
                    for k, v in kwargs.items()
                }
                return fake_func(*args, **kwargs)
            
            _lib.impl(op_name, wrapper, "Meta")
            return fake_func
        return decorator
else:
    # PyTorch 2.2+ 使用 register_fake 或 impl_abstract
    try:
        from torch.library import register_fake as _native_register_fake
    except ImportError:
        from torch.library import impl_abstract as _native_register_fake
    
    def _compatible_register_fake(op_name: str):
        """Compatibility wrapper for PyTorch 2.2+ fake registration."""
        return _native_register_fake(op_name)



def register_mindie_fake_op(op_name: str):
    """Decorator to register a fake implementation for a MindIE operator.
    
    Usage:
        @register_mindie_fake_op("rope")
        def rope_fake(x, cos, sin, mode):
            ...
    
    Args:
        op_name: Full name of the operator (e.g. "rope", "la")
    
    Returns:
        Decorator function that registers the fake implementation.
    """
    if not is_npu_available():
        def dummy_decorator(func):
            return func
        return dummy_decorator
    
    if not check_mindie_operator_exists(op_name):
        raise RuntimeError(
            f"MindIE operator {MINDIE_NS}::{op_name} not found! "
            "Ensure the SO library is loaded and the operator is registered with TORCH_LIBRARY."
        )
    
    return _compatible_register_fake(f"{MINDIE_NS}::{op_name}")